File: 10. Dynamic Programming/09. Best time to buy and sell stock with transaction fee/main.java
import java.util.Arrays;

public class main {

    // 1. Brute Force (Recursive) - Exponential Time Complexity (O(2^N))
    public static int bruteForce(int[] prices, int i, boolean holding, int fee) {
        if (i >= prices.length) return 0;

        if (holding) {
            return Math.max(prices[i] - fee + bruteForce(prices, i + 1, false, fee),
                            bruteForce(prices, i + 1, true, fee));
        } else {
            return Math.max(-prices[i] + bruteForce(prices, i + 1, true, fee),
                            bruteForce(prices, i + 1, false, fee));
        }
    }

    public static int maxProfitBruteForce(int[] prices, int fee) {
        return bruteForce(prices, 0, false, fee);
    }

    // 2. Recursion + Memoization (O(N) Time, O(N) Space)
    public static int memoizedDFS(int[] prices, int i, boolean holding, int fee, int[][] memo) {
        if (i >= prices.length) return 0;
        if (memo[i][holding ? 1 : 0] != -1) return memo[i][holding ? 1 : 0];

        if (holding) {
            return memo[i][1] = Math.max(prices[i] - fee + memoizedDFS(prices, i + 1, false, fee, memo),
                                         memoizedDFS(prices, i + 1, true, fee, memo));
        } else {
            return memo[i][0] = Math.max(-prices[i] + memoizedDFS(prices, i + 1, true, fee, memo),
                                         memoizedDFS(prices, i + 1, false, fee, memo));
        }
    }

    public static int maxProfitMemo(int[] prices, int fee) {
        int n = prices.length;
        int[][] memo = new int[n][2];
        for (int[] row : memo) Arrays.fill(row, -1);
        return memoizedDFS(prices, 0, false, fee, memo);
    }

    // 3. Bottom-Up Dynamic Programming (O(N) Time, O(N) Space)
    public static int maxProfitDP(int[] prices, int fee) {
        int n = prices.length;
        int[][] dp = new int[n + 1][2];

        for (int i = n - 1; i >= 0; i--) {
            dp[i][1] = Math.max(prices[i] - fee + dp[i + 1][0], dp[i + 1][1]); // Holding
            dp[i][0] = Math.max(-prices[i] + dp[i + 1][1], dp[i + 1][0]); // Not holding
        }
        return dp[0][0];
    }

    // 4. Optimized DP with O(1) Space (O(N) Time, O(1) Space)
    public static int maxProfitOptimizedDP(int[] prices, int fee) {
        int aheadNotHolding = 0, aheadHolding = 0;

        for (int i = prices.length - 1; i >= 0; i--) {
            int tempNotHolding = Math.max(-prices[i] + aheadHolding, aheadNotHolding);
            int tempHolding = Math.max(prices[i] - fee + aheadNotHolding, aheadHolding);
            aheadNotHolding = tempNotHolding;
            aheadHolding = tempHolding;
        }
        return aheadNotHolding;
    }

    // 5. Greedy Approach (O(N) Time, O(1) Space) - Best Solution
    public static int maxProfitGreedy(int[] prices, int fee) {
        int profit = 0;
        int minPrice = prices[0];

        for (int i = 1; i < prices.length; i++) {
            if (prices[i] < minPrice) {
                minPrice = prices[i]; // Update minimum buying price
            }
            if (prices[i] > minPrice + fee) {
                profit += prices[i] - minPrice - fee; // Sell stock and add profit
                minPrice = prices[i] - fee; // Adjust minPrice for next transaction
            }
        }
        return profit;
    }

    // Main function to test different approaches
    public static void main(String[] args) {
        int[] prices = {1, 3, 2, 8, 4, 9}; // Example input
        int fee = 2;

        System.out.println("Brute Force: " + maxProfitBruteForce(prices, fee));
        System.out.println("Memoization: " + maxProfitMemo(prices, fee));
        System.out.println("Bottom-Up DP: " + maxProfitDP(prices, fee));
        System.out.println("Optimized DP: " + maxProfitOptimizedDP(prices, fee));
        System.out.println("Greedy Approach: " + maxProfitGreedy(prices, fee));
    }
}

File: 10. Dynamic Programming/09. Best time to buy and sell stock with transaction fee/main.go
package main

import (
	"fmt"
)

// 1. Brute Force (Recursive)
func bruteForce(prices []int, i int, holding bool, fee int) int {
	if i >= len(prices) {
		return 0
	}
	if holding {
		return max(
			prices[i]-fee+bruteForce(prices, i+1, false, fee), // Sell
			bruteForce(prices, i+1, true, fee),               // Hold
		)
	} else {
		return max(
			-prices[i]+bruteForce(prices, i+1, true, fee), // Buy
			bruteForce(prices, i+1, false, fee),           // Skip
		)
	}
}

func maxProfitBruteForce(prices []int, fee int) int {
	return bruteForce(prices, 0, false, fee)
}

// 2. Memoization
func memoizedDFS(prices []int, i int, holding int, fee int, memo [][]int) int {
	if i >= len(prices) {
		return 0
	}
	if memo[i][holding] != -1 {
		return memo[i][holding]
	}

	if holding == 1 {
		memo[i][holding] = max(
			prices[i]-fee+memoizedDFS(prices, i+1, 0, fee, memo), // Sell
			memoizedDFS(prices, i+1, 1, fee, memo),               // Hold
		)
	} else {
		memo[i][holding] = max(
			-prices[i]+memoizedDFS(prices, i+1, 1, fee, memo), // Buy
			memoizedDFS(prices, i+1, 0, fee, memo),            // Skip
		)
	}
	return memo[i][holding]
}

func maxProfitMemo(prices []int, fee int) int {
	n := len(prices)
	memo := make([][]int, n)
	for i := range memo {
		memo[i] = []int{-1, -1}
	}
	return memoizedDFS(prices, 0, 0, fee, memo)
}

// 3. Bottom-Up DP
func maxProfitDP(prices []int, fee int) int {
	n := len(prices)
	if n == 0 {
		return 0
	}
	dp := make([][]int, n+1)
	for i := range dp {
		dp[i] = []int{0, 0}
	}
	for i := n - 1; i >= 0; i-- {
		dp[i][1] = max(prices[i]-fee+dp[i+1][0], dp[i+1][1])
		dp[i][0] = max(-prices[i]+dp[i+1][1], dp[i+1][0])
	}
	return dp[0][0]
}

// 4. Optimized DP
func maxProfitOptimizedDP(prices []int, fee int) int {
	n := len(prices)
	if n == 0 {
		return 0
	}
	aheadNotHolding, aheadHolding := 0, 0
	for i := n - 1; i >= 0; i-- {
		tempNotHolding := max(-prices[i]+aheadHolding, aheadNotHolding)
		tempHolding := max(prices[i]-fee+aheadNotHolding, aheadHolding)
		aheadNotHolding = tempNotHolding
		aheadHolding = tempHolding
	}
	return aheadNotHolding
}

// 5. Greedy Approach
func maxProfitGreedy(prices []int, fee int) int {
	if len(prices) == 0 {
		return 0
	}
	profit := 0
	minPrice := prices[0]
	for i := 1; i < len(prices); i++ {
		if prices[i] < minPrice {
			minPrice = prices[i]
		}
		if prices[i] > minPrice+fee {
			profit += prices[i] - minPrice - fee
			minPrice = prices[i] - fee
		}
	}
	return profit
}

// Utility
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// Main
func main() {
	prices := []int{1, 3, 2, 8, 4, 9}
	fee := 2

	fmt.Println("Brute Force:", maxProfitBruteForce(prices, fee))
	fmt.Println("Memoization:", maxProfitMemo(prices, fee))
	fmt.Println("Bottom-Up DP:", maxProfitDP(prices, fee))
	fmt.Println("Optimized DP:", maxProfitOptimizedDP(prices, fee))
	fmt.Println("Greedy Approach:", maxProfitGreedy(prices, fee))
}
